bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/services/amp/AmpTransformer.java

```diff
@@ -97,6 +139,23 @@ public void startElement(String uri, String localName, String qName, Attributes
         contentHandler.startElement(uri, localName, qName, atts);
     }
 
+    @Override
+    public void characters(char[] ch, int start, int length) throws SAXException {
+
+        // Append the aggregated js to the head.
+        if (!jsAppended && isAmpMode && isAmpSelector) {
+            String content = new String(ch);
+            if (content.contains(HEAD_TAG)) {
```

Comment:
I wouldn't trust this - i don't know the rhyme or reason behind the way the `characters` breaks up the HTML doc into bits to be processed, but having this check here leaves the potential that the document gets processed in chunks with `<hea` in one iteration and `d>` in the next, so in that output your JS would be skipped over. Check out this doc on 'Character Events' in the SAX XML parser: https://docs.oracle.com/javase/tutorial/jaxp/sax/parsing.html#gclmx
Important bit:

> Parsers are not required to return any particular number of characters at one time. A parser can return anything from a single character at a time up to several thousand and still be a standard-conforming implementation. So if your application needs to process the characters it sees, it is wise to have the characters() method accumulate the characters in a java.lang.StringBuffer and operate on them only when you are sure that all of them have been found.

Reply:
wow, nice catch